### PR TITLE
MAINT: use mixin for classes storing metadata

### PR DESCRIPTION
**Pull request by jairideout**
We were duplicating code on ``Sequence`` and ``TabularMSA`` to support arbitrary metadata. Most, if not all, classes in scikit-bio will eventually support metadata so this refactoring will avoid further code duplication as the project grows.

This commit adds a private mixin ``skbio._base.MetadataMixin`` which centralizes the following metadata logic:

- `metadata` property (getter, setter, deleter)
- `has_metadata` method
- metadata handling in `__init__`
- metadata handling in `__eq__`/`__ne__`

Additionally, a private unit test mixin `skbio.util._testing.MetadataMixinTests` is used to test `MetadataMixin`, `Sequence`, and `TabularMSA`, avoiding redundant unit tests and ensuring that metadata API and behavior is consistent across classes.

### CODE REVIEW

**Comment by anderspitman** [comment]
:+1: I was just thinking about this today. Thanks for picking up the brainwaves and implementing it!

File: skbio/sequence/_sequence.py [hunk 1]
```diff
@@ -887,18 +808,7 @@ def __eq__(self, other):
...
-        # positional metadata
-        if self.has_metadata() and other.has_metadata():
-            if self.metadata != other.metadata:
-                return False
-        elif not (self.has_metadata() or other.has_metadata()):
-            # both don't have metadata
-            pass
-        else:
-            # one has metadata while the other does not
+        if not super(Sequence, self).__eq__(other):
```
Comment (anderspitman) [inline skbio/sequence/_sequence.py]: Nice! I was wondering what a good way to do this without explicitly depending on the implementation of the mixin would be.

**Comment by jairideout** [comment]
No problem, I was starting to work on adding positional metadata to TabularMSA and couldn't bring myself to copy and paste again ;)

**Comment by ebolyen** [comment]
@jairideout looks like we have some merge conflicts

**Comment by jairideout** [comment]
Noooooooooooooo

**Comment by jairideout** [comment]
@ebolyen merge conflicts are resolved

**Comment by ebolyen** [comment]
:+1: :x: :100: 

This is awesome!

**Pull request merged**